# Lab2
#### Fidel Diaz | 3-751-509 | 1SF125 | fidel.diaz@utp.ac.pa
#### Sebastian Mejia | 4-816-1428 | 1SF125 | sebastian.mejia@utp.ac.pa

# Problem1

## Best Case

Realice un script que implemente las 3 versiones arriba descritas para el problema de
elementos comunes en dos arreglos.

In [4]:
#Algorithm1 Versión Sencilla (Naive) 
import math as mt
def search1(a,n,b,m):
    i, j, intersect = 0,0,0
    for i in range(n):
        for j in range(m):
            if a[i] == b[j]:
                intersect = intersect + 1
                break
    return intersect
#Algorithm2 Version con Busqueda(simple)
def search2(a,n,b,m):
    i, intersect = 0, 0
    for i in range(n):
        if searched(b, m, a[i]):
            intersect = intersect + 1
    return intersect
def searched(b,m,val):
    i = 0
    while(i < m):
        if(b[i] == val):
            return True
        i = i + 1
    return False
#Algorithm3 Versión con Búsqueda Binaria
def search3(a,n,b,m):
    i, intersect = 0, 0
    for i in range(n):
        if binarysearch(b, m, a[i]):
            intersect = intersect + 1
    return intersect
def binarysearch(b, m, val):
    l = 0
    r = m - 1
    while(r >= l):
        mid = mt.floor((l+r)/2)
        if(b[mid] == val):
            return True
        else:
            if(b[mid] < val):
                l = mid + 1
            else:
                r = mid - 1
    return False
#main function
def main():
    #Best Case
    A = list(range(11))
    B = list(range(11))
    # Use set() to eliminate duplicates
    setA = set(A)
    setB = set(B)
    A.sort()
    B.sort()
    A = list(setA)
    B = list(setB)
    n = len(A)
    m = len(B)
    x1 = search1(A, n, B, m)
    x2 = search2(A, n, B, m)
    x3 = search3(A, n, B, m)
    print("Algorithm1 | Quantity of items in A and B lists:", x1)
    print("Algorithm2 | Quantity of items in A and B lists:", x2)
    print("Algorithm3 | Quantity of items in A and B lists:", x3)
if __name__ == "__main__":
    main()

Algorithm1 | Quantity of items in A and B lists: 11
Algorithm2 | Quantity of items in A and B lists: 11
Algorithm3 | Quantity of items in A and B lists: 11


# Problem2

## Best Case | Execution time of Algorithm

### Using time module.

Agregar tabla con entrada n, m tiempo de ejecucion y solo 2 algoritmos.

In [3]:
import time as t
import math as mt
# Algorithm1 Versión Sencilla (Naive)
def search1(a, n, b, m):
    i, j, intersect = 0, 0, 0
    for i in range(n):
        for j in range(m):
            if a[i] == b[j]:
                intersect = intersect + 1
                break
    return intersect
# Algorithm2 Version con Busqueda(simple)
def search2(a, n, b, m):
    i, intersect = 0, 0
    for i in range(n):
        if searched(b, m, a[i]):
            intersect = intersect + 1
    return intersect
def searched(b, m, val):
    i = 0
    while(i < m):
        if(b[i] == val):
            return True
        i = i + 1
    return False
# Algorithm3 Versión con Búsqueda Binaria
def search3(a, n, b, m):
    i, intersect = 0, 0
    for i in range(n):
        if binarysearch(b, m, a[i]):
            intersect = intersect + 1
    return intersect
def binarysearch(b, m, val):
    l = 0
    r = m - 1
    while(r >= l):
        mid = mt.floor((l+r)/2)
        if(b[mid] == val):
            return True
        else:
            if(b[mid] < val):
                l = mid + 1
            else:
                r = mid - 1
    return False
if __name__ == "__main__":
    #Best Case
    A = list(range(11))
    B = list(range(11))
    # Use set() to eliminate duplicates
    setA = set(A)
    setB = set(B)
    A = list(setA)
    B = list(setB)
    A.sort()
    B.sort()
    n = len(A)
    m = len(B)
    # Measure execution time for Algorithm1
    t1 = t.time()
    x1 = search1(A, n, B, m)
    t2 = t.time()
    execution_time_1 = t2 - t1
    print("Execution time of Algorithm1:", execution_time_1, "seconds")
    # Measure execution time for Algorithm2
    t1 = t.time()
    x2 = search2(A, n, B, m)
    t2 = t.time()
    execution_time_2 = t2 - t1
    print("Execution time of Algorithm2:", execution_time_2, "seconds")
    # Measure execution time for Algorithm3
    t1 = t.time()
    x3 = search3(A, n, B, m)
    t2 = t.time()
    execution_time_3 = t2 - t1
    print("Execution time of Algorithm3:", execution_time_3, "seconds")

Execution time of Algorithm1: 0.0 seconds
Execution time of Algorithm2: 0.0 seconds
Execution time of Algorithm3: 0.0 seconds


### Using timeit module

In [4]:
import timeit as ti
import math as mt
# Algorithm1 Versión Sencilla (Naive)
def search1(a, n, b, m):
    i, j, intersect = 0, 0, 0
    for i in range(n):
        for j in range(m):
            if a[i] == b[j]:
                intersect = intersect + 1
                break
    return intersect

# Algorithm2 Version con Busqueda(simple)
def search2(a, n, b, m):
    i, intersect = 0, 0
    for i in range(n):
        if searched(b, m, a[i]):
            intersect = intersect + 1
    return intersect

def searched(b, m, val):
    i = 0
    while(i < m):
        if(b[i] == val):
            return True
        i = i + 1
    return False

# Algorithm3 Versión con Búsqueda Binaria
def search3(a, n, b, m):
    i, intersect = 0, 0
    for i in range(n):
        if binarysearch(b, m, a[i]):
            intersect = intersect + 1
    return intersect

def binarysearch(b, m, val):
    l = 0
    r = m - 1
    while(r >= l):
        mid = mt.floor((l+r)/2)
        if(b[mid] == val):
            return True
        else:
            if(b[mid] < val):
                l = mid + 1
            else:
                r = mid - 1
    return False

# Measure execution time using timeit library
if __name__ == "__main__":
    #Best Case
    A = list(range(11))
    B = list(range(11))
    # Use set() to eliminate duplicates
    setA = set(A)
    setB = set(B)
    A = list(setA)
    B = list(setB)
    A.sort()
    B.sort()
    n = len(A)
    m = len(B)

    # Measure execution time for Algorithm1 with timeit
    execution_time_1 = ti.timeit("search1(A, n, B, m)", globals=globals(), number=10000)
    print("Execution time of Algorithm1:", execution_time_1, "seconds")

    # Measure execution time for Algorithm2 with timeit
    execution_time_2 = ti.timeit("search2(A, n, B, m)", globals=globals(), number=10000)
    print("Execution time of Algorithm2:", execution_time_2, "seconds")

    # Measure execution time for Algorithm3 with timeit
    execution_time_3 = ti.timeit("search3(A, n, B, m)", globals=globals(), number=10000)
    print("Execution time of Algorithm3:", execution_time_3, "seconds")

Execution time of Algorithm1: 0.08379730000160635 seconds
Execution time of Algorithm2: 0.08190249999825028 seconds
Execution time of Algorithm3: 0.13883760000317125 seconds


## Worst Case | Execution time of Algorithms

Without sort()

In [12]:
import pandas as pd
import timeit as ti
import random

# Function to generate two lists a and b of size (n, m) with no common elements
def generate_lists(size):
    n = 2 ** size
    a = random.sample(range(1, n * 10), n)
    b = random.sample(range(n * 10 + 1, (n + n) * 10), n)
    seta = set(a)
    setb = set(b)
    a = list(seta)
    b = list(setb)
    a.sort()
    b.sort()
    return a, b

# Algorithm1 Versión Sencilla (Naive)
def search1(a, n, b, m):
    intersect = 0
    for i in range(n):
        for j in range(m):
            if a[i] == b[j]:
                intersect = intersect + 1
                break
    return intersect

# Algorithm3 Versión con Búsqueda Binaria
def search3(a, n, b, m):
    intersect = 0
    for i in range(n):
        if binarysearch(b, m, a[i]):
            intersect = intersect + 1
    return intersect

def binarysearch(b, m, val):
    l = 0
    r = m - 1
    while(r >= l):
        mid = (l + r) // 2
        if(b[mid] == val):
            return True
        else:
            if(b[mid] < val):
                l = mid + 1
            else:
                r = mid - 1
    return False

if __name__ == "__main__":
    results = {
        '(n, m)': [],
        'Algorithm1': [],
        'Algorithm3': []
    }
    for size in range(3, 12):
        a, b = generate_lists(size)
        n = len(a)
        m = len(b)

        # Measure execution time for Algorithm1 with timeit
        execution_time_1 = ti.timeit("search1(a, n, b, m)", globals=globals(), number=100)
        
        # Measure execution time for Algorithm3 with timeit
        execution_time_3 = ti.timeit("search3(a, n, b, m)", globals=globals(), number=100)

        results['(n, m)'].append((n, m))
        results['Algorithm1'].append(execution_time_1)
        results['Algorithm3'].append(execution_time_3)

    # Create a Pandas DataFrame
    df = pd.DataFrame(results)

    # Print the DataFrame
    print(df)

         (n, m)  Algorithm1  Algorithm3
0        (8, 8)    0.000964    0.001041
1      (16, 16)    0.002773    0.002327
2      (32, 32)    0.006247    0.005081
3      (64, 64)    0.033290    0.010186
4    (128, 128)    0.210970    0.049597
5    (256, 256)    0.470406    0.055410
6    (512, 512)    1.972668    0.160673
7  (1024, 1024)    8.945198    0.302252
8  (2048, 2048)   32.924080    0.685536


With sort()

In [14]:
import pandas as pd
import timeit as ti
import random

# Función para generar dos listas a y b de tamaño (n, m) sin elementos en común
def generate_lists(n):
    n = 2 ** size
    a = random.sample(range(1, n * 10), n)
    b = random.sample(range(n * 10 + 1, (n + n) * 10), n)
    return a, b

# Algorithm1 Versión Sencilla (Naive)
def search1(a, n, b, m):
    intersect = 0
    for i in range(n):
        for j in range(m):
            if a[i] == b[j]:
                intersect = intersect + 1
                break
    return intersect

# Algorithm3 Versión con Búsqueda Binaria
def search3(a, n, b, m):
    intersect = 0
    a.sort()
    b.sort()
    for i in range(n):
        if binarysearch(b, m, a[i]):
            intersect = intersect + 1
    return intersect

def binarysearch(b, m, val):
    l = 0
    r = m - 1
    while(r >= l):
        mid = (l + r) // 2
        if(b[mid] == val):
            return True
        else:
            if(b[mid] < val):
                l = mid + 1
            else:
                r = mid - 1
    return False

if __name__ == "__main__":
    results = {
        '(n, m)': [],
        'Algorithm1': [],
        'Algorithm3': []
    }

    for size in range(3, 12):
        a, b = generate_lists(size)
        n = len(a)
        m = len(b)

        # Measure execution time for Algorithm1 with timeit
        execution_time_1 = ti.timeit("search1(a, n, b, m)", globals=globals(), number=100)
        
        # Measure execution time for Algorithm3 with timeit
        execution_time_3 = ti.timeit("search3(a, n, b, m)", globals=globals(), number=100)

        results['(n, m)'].append((n, m))
        results['Algorithm1'].append(execution_time_1)
        results['Algorithm3'].append(execution_time_3)

    # Create a Pandas DataFrame
    df = pd.DataFrame(results)

    # Print the DataFrame
    print(df)

         (n, m)  Algorithm1  Algorithm3
0        (8, 8)    0.000628    0.000793
1      (16, 16)    0.004178    0.002777
2      (32, 32)    0.007855    0.003846
3      (64, 64)    0.029721    0.010321
4    (128, 128)    0.130660    0.027153
5    (256, 256)    0.584529    0.049087
6    (512, 512)    2.446874    0.143932
7  (1024, 1024)    8.012869    0.279721
8  (2048, 2048)   33.266959    0.663914


# Problem3

##### Realice un script que implemente la versión recursiva de la búsqueda binaria para el problema de elementos comunes en dos arreglos. Utilizando el algoritmo dado en clase.

In [22]:
#Algorithm4 Versión con Búsqueda Binaria Recursiva
import math as mt
def search4(a,n,b,m):
    i, intersect = 0, 0
    for i in range(n):
        if recursivebinarysearch(b, 0, m - 1, a[i]):
            intersect = intersect + 1
    return intersect
def recursivebinarysearch(b, l, r,val):
    # Check base case
    if(l > r):
        return False
    mid = mt.floor((l+r)/2)
    # If element is present at the middle itself
    if(b[mid] == val):
        return True
    else:
        if(b[mid] < val):
            l = mid + 1
            return recursivebinarysearch(b, l, r, val)
        else:
            r = mid - 1
            return recursivebinarysearch(b, l, r, val)
#main function
if __name__ == "__main__":
    #Best Case
    A = list(range(11))
    B = list(range(11))
    # Use set() to eliminate duplicates
    setA = set(A)
    setB = set(B)
    A = list(setA)
    B = list(setB)
    A.sort()
    B.sort()
    n = len(A)
    m = len(B)
    x4 = search4(A, n, B, m)
    print("Algorithm4 | Quantity of items in A and B lists:", x4)

Algorithm4 | Quantity of items in A and B lists: 11


## Best Case | Execution time of Algorithm

In [18]:
#Algorithm4 Versión con Búsqueda Binaria Recursiva
import math as mt
def search4(a,n,b,m):
    i, intersect = 0, 0
    for i in range(n):
        if recursivebinarysearch(b, 0, m - 1, a[i]):
            intersect = intersect + 1
    return intersect
def recursivebinarysearch(b, l, r,val):
    # Check base case
    if(l > r):
        return False
    mid = mt.floor((l+r)/2)
    # If element is present at the middle itself
    if(b[mid] == val):
        return True
    else:
        if(b[mid] < val):
            l = mid + 1
            return recursivebinarysearch(b, l, r, val)
        else:
            r = mid - 1
            return recursivebinarysearch(b, l, r, val)
#main function
if __name__ == "__main__":
    #Best Case
    A = list(range(11))
    B = list(range(11))
    # Use set() to eliminate duplicates
    setA = set(A)
    setB = set(B)
    A = list(setA)
    B = list(setB)
    A.sort()
    B.sort()
    n = len(A)
    m = len(B)
    execution_time_4 = ti.timeit("search4(A, n, B, m)", globals=globals(), number=10000)
    print("Execution time of Algorithm3:", execution_time_4, "seconds")

Execution time of Algorithm3: 0.17341200000373647 seconds


## Worst Case | Execution time of Algorithm

In [21]:
#Algorithm4 Versión con Búsqueda Binaria Recursiva
import math as mt
def search4(a,n,b,m):
    i, intersect = 0, 0
    for i in range(n):
        if recursivebinarysearch(b, 0, m - 1, a[i]):
            intersect = intersect + 1
    return intersect
def recursivebinarysearch(b, l, r,val):
    # Check base case
    if(l > r):
        return False
    mid = mt.floor((l+r)/2)
    # If element is present at the middle itself
    if(b[mid] == val):
        return True
    else:
        if(b[mid] < val):
            l = mid + 1
            return recursivebinarysearch(b, l, r, val)
        else:
            r = mid - 1
            return recursivebinarysearch(b, l, r, val)
#main function
if __name__ == "__main__":
    #Best Case
    A = list(range(11))
    C = list(range(11, 22))
    # Use set() to eliminate duplicates
    setA = set(A)
    setC = set(C)
    A = list(setA)
    C = list(setC)
    A.sort()
    C.sort()
    n = len(A)
    m = len(C)
    execution_time_4 = ti.timeit("search4(A, n, C, m)", globals=globals(), number=10000)
    print("Execution time of Algorithm3:", execution_time_4, "seconds")

Execution time of Algorithm3: 0.15300079999724403 seconds


## Worst Case | Execution time of Algorithms

In [25]:
import pandas as pd
import timeit as ti
import random
import math as mt

# Function to generate two lists a and b of size (n, m) with no common elements
def generate_lists(size):
    n = 2 ** size
    a = random.sample(range(1, n * 10), n)
    b = random.sample(range(n * 10 + 1, (n + n) * 10), n)
    seta = set(a)
    setb = set(b)
    a = list(seta)
    b = list(setb)
    a.sort()
    b.sort()
    return a, b

# Algorithm1 Versión Sencilla (Naive)
def search1(a, n, b, m):
    intersect = 0
    for i in range(n):
        for j in range(m):
            if a[i] == b[j]:
                intersect = intersect + 1
                break
    return intersect

# Algorithm2 Version con Búsqueda Simple
def search2(a, n, b, m):
    intersect = 0
    for i in range(n):
        if a[i] in b:
            intersect = intersect + 1
    return intersect

# Algorithm3 Versión con Búsqueda Binaria
def search3(a, n, b, m):
    intersect = 0
    for i in range(n):
        if binarysearch(b, m, a[i]):
            intersect = intersect + 1
    return intersect

def binarysearch(b, m, val):
    l = 0
    r = m - 1
    while(r >= l):
        mid = (l + r) // 2
        if(b[mid] == val):
            return True
        else:
            if(b[mid] < val):
                l = mid + 1
            else:
                r = mid - 1
    return False

# Algorithm4 Versión con Búsqueda Binaria Recursiva
def search4(a, n, b, m):
    intersect = 0
    for i in range(n):
        if recursivebinarysearch(b, 0, m - 1, a[i]):
            intersect = intersect + 1
    return intersect

def recursivebinarysearch(b, l, r, val):
    if l > r:
        return False
    mid = mt.floor((l + r) / 2)
    if b[mid] == val:
        return True
    elif b[mid] < val:
        l = mid + 1
        return recursivebinarysearch(b, l, r, val)
    else:
        r = mid - 1
        return recursivebinarysearch(b, l, r, val)

if __name__ == "__main__":
    results = {
        '(n, m)': [],
        'Algorithm1': [],
        'Algorithm2': [],
        'Algorithm3': [],
        'Algorithm4': []
    }
    for size in range(3, 12):
        a, b = generate_lists(size)
        n = len(a)
        m = len(b)

        # Measure execution time for Algorithm1 with timeit
        execution_time_1 = ti.timeit("search1(a, n, b, m)", globals=globals(), number=100)
        
        # Measure execution time for Algorithm2 with timeit
        execution_time_2 = ti.timeit("search2(a, n, b, m)", globals=globals(), number=100)

        # Measure execution time for Algorithm3 with timeit
        execution_time_3 = ti.timeit("search3(a, n, b, m)", globals=globals(), number=100)

        # Measure execution time for Algorithm4 with timeit
        execution_time_4 = ti.timeit("search4(a, n, b, m)", globals=globals(), number=100)

        results['(n, m)'].append((n, m))
        results['Algorithm1'].append(execution_time_1)
        results['Algorithm2'].append(execution_time_2)
        results['Algorithm3'].append(execution_time_3)
        results['Algorithm4'].append(execution_time_4)

    # Create a Pandas DataFrame
    df = pd.DataFrame(results)

    # Print the DataFrame
    print(df)

    # Print algorithm names
    algorithm_names = {
        'Algorithm1': 'Algorithm1 Versión Sencilla (Naive)',
        'Algorithm2': 'Algorithm2 Version con Búsqueda Simple',
        'Algorithm3': 'Algorithm3 Versión con Búsqueda Binaria',
        'Algorithm4': 'Algorithm4 Versión con Búsqueda Binaria Recursiva'
    }

    for algorithm in algorithm_names.values():
        print(algorithm)

         (n, m)  Algorithm1  Algorithm2  Algorithm3  Algorithm4
0        (8, 8)    0.000708    0.000182    0.000761    0.001105
1      (16, 16)    0.002188    0.000339    0.001857    0.002778
2      (32, 32)    0.005966    0.001059    0.003369    0.005350
3      (64, 64)    0.023456    0.003521    0.008846    0.015306
4    (128, 128)    0.112074    0.018419    0.020430    0.028293
5    (256, 256)    0.299897    0.055133    0.041442    0.063935
6    (512, 512)    1.323979    0.295609    0.099736    0.158232
7  (1024, 1024)    5.677106    1.144717    0.233569    0.364022
8  (2048, 2048)   26.170322    4.681131    0.556523    0.887102
Algorithm1 Versión Sencilla (Naive)
Algorithm2 Version con Búsqueda Simple
Algorithm3 Versión con Búsqueda Binaria
Algorithm4 Versión con Búsqueda Binaria Recursiva


## Questions
##### 1. Como se comporta este algoritmo cuando existe al menos un traslape entre los arreglos A y B?
##### Respuesta: El tiempo de ejecucion en el mejor caso, el Algoritmo4 es mas rapido que el algoritmo 1 y 2.
##### 2. como se comporta este algoritmo en el peor caso cuando es comparado contra el mejor de los otros?
##### Execution time of Algorithm4: 0.001105
##### Execution time of Algorithm1: 0.08379730000160635 seconds
##### Execution time of Algorithm2: 0.08190249999825028 seconds
##### Execution time of Algorithm3: 0.13883760000317125 seconds
##### Respuesta: El tiempo de ejecucion en el peor caso del Algoritmo4 tiende a ser mas rapido que el mejor caso de los otros algoritmos.
## Observaciones
##### 1. La libreria time no se comporta igual que timeit en todos los IDES.
##### 2. Los algoritmos difieren en tiempo de ejecución si dentro de ellos se incluye un sort().
##### 3. Existe diferencua entre solo usar if __name__ == "__main__": y def main:
##### 4. Existe diferencia en el tiempo de ejecucion dependiendo si las listas tienen todos sus elementos comunes, si tienen ciertos elementos comunes y si no tienen ninguno. 
##### 4. Ejemplo: A = [1, 2, 3] B = [1, 2, 3] es diferete de A = [1, 2, 3] B = [4, 1, 6] y diferente de  A = [1, 2, 3] B = [4, 5, 6]. Existe una gran diferente al momento de determinar el mejor caso, caso promedio y peor caso.

##### Sugerencia: Por favor resolver las dudas en la proximas clases, especiamente la observacion 4.